src/pki/commands/ocsp.c

```diff
@@ -0,0 +1,569 @@
+/*
+ * Copyright (C) 2023 Andreas Steffen, strongSec GmbH
+ *
+ * Copyright (C) secunet Security Networks AG
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation; either version 2 of the License, or (at your
+ * option) any later version.  See <http://www.fsf.org/copyleft/gpl.txt>.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * for more details.
+ */
+
+#include <errno.h>
+#include <time.h>
+
+#include "pki.h"
+
+#include <credentials/sets/mem_cred.h>
+#include <credentials/certificates/ocsp_request.h>
+#include <credentials/certificates/ocsp_response.h>
+#include <credentials/certificates/ocsp_single_response.h>
+#include <credentials/certificates/ocsp_responder.h>
+
+static bool verify_signature(certificate_t *ocsp_req, mem_cred_t *creds)
+{
+	identification_t *signer;
+	certificate_t *signer_cert, *cert_found;
+	ocsp_request_t *ocsp_request;
+	enumerator_t *certs;
+	bool trusted = TRUE;
+
+	ocsp_request = (ocsp_request_t*)ocsp_req;
+
+	signer_cert = ocsp_request->get_signer_cert(ocsp_request);
+	if (signer_cert)
+	{
+		signer = signer_cert->get_subject(signer_cert);
+
+		/* establish trust relative to root CA */
+		creds->add_cert(creds, FALSE, signer_cert->get_ref(signer_cert));
+		certs = lib->credmgr->create_trusted_enumerator(lib->credmgr,
+											KEY_ANY, signer, FALSE);
+		trusted = certs->enumerate(certs, &cert_found, NULL) &&
+								  (cert_found == signer_cert);
+		certs->destroy(certs);
+		trusted &= ocsp_req->issued_by(ocsp_req, signer_cert, NULL);
+	}
+	DBG1(DBG_LIB, "requestor is %strusted", trusted ? "" : "not ");
+
+	return trusted;
+}
+
+static bool find_issuer_cacert(hash_algorithm_t hashAlgorithm,
+							   chunk_t issuerKeyHash, chunk_t issuerNameHash,
+							   certificate_t **issuer_cacert)
+{
+    bool issuerKeyHash_ok = FALSE, issuerNameHash_ok = FALSE;
+	certificate_t *candidate;
+	identification_t *issuer;
+	x509_t *x509_ca;
+	chunk_t caKeyHash = chunk_empty, caNameHash = chunk_empty;
+	hasher_t *hasher;
+	enumerator_t *certs;
+
+	*issuer_cacert = NULL;
+
+	certs = lib->credmgr->create_cert_enumerator(lib->credmgr, CERT_X509,
+												 KEY_ANY, NULL, TRUE);
+	while (certs->enumerate(certs, &candidate) && hashAlgorithm == HASH_SHA1)
+	{
+		x509_ca = (x509_t*)candidate;
+
+		if (!(x509_ca->get_flags(x509_ca) & X509_CA))
+		{
+			continue;
+		}
+
+		/* retrieve the caKeyHash of the candidate issuer */
+		caKeyHash = x509_ca->get_subjectKeyIdentifier(x509_ca);
+		issuerKeyHash_ok = chunk_equals(issuerKeyHash, caKeyHash);
```

Comment:
What does **`chunk_equals_const()`** protect against?

Reply:
Just thought because it will be called remotely and may allow an attacker to determine for which CA the OCSP server is responsible (in case that's not obvious I guess).